vm/vmm/vmm.go

```diff
@@ -0,0 +1,422 @@
+	signal := func(err error) {
+		select {
+		case errc <- err:
+		default:
+		}
+	}
+
+	go func() {
+		select {
+		case <-time.After(timeout):
+			signal(vmimpl.ErrTimeout)
+		case <-stop:
+			signal(vmimpl.ErrTimeout)
+		case err := <-inst.merger.Err:
+			cmd.Process.Kill()
+			if cmdErr := cmd.Wait(); cmdErr == nil {
+				// If the command exited successfully, we got EOF error from merger.
+				// But in this case no error has happened and the EOF is expected.
+				err = nil
+			}
+
+			signal(err)
+			return
+		}
+		cmd.Process.Kill()
+		cmd.Wait()
+	}()
+	return inst.merger.Output, errc, nil
+}
+
+func (inst *instance) Diagnose() bool {
+	didWrite := false
+	commands := []string{"", "trace", "show registers"}
+	for _, c := range commands {
+		if inst.wpipe != nil {
+			didWrite = true
+			inst.wpipe.Write([]byte(c + "\n"))
+			time.Sleep(100 * time.Millisecond)
+		}
+	}
+	return didWrite
+}
+
+func (inst *instance) console() {
+	run := func() error {
+		outr, outw, err := osutil.LongPipe()
+		if err != nil {
+			return fmt.Errorf("failed to create output pipe: %v", err)
+		}
+		inr, inw, err := osutil.LongPipe()
+		if err != nil {
+			return fmt.Errorf("failed to create input pipe: %v", err)
+		}
+		inst.wpipe = inw
+
+		cmd := osutil.Command("vmctl", "console", inst.vmIdent())
+		cmd.Stdin = inr
+		cmd.Stdout = outw
+		cmd.Stderr = outw
+		if err := cmd.Start(); err != nil {
+			return err
+		}
+		outw.Close()
+		inr.Close()
+		inst.merger.Add("console", outr)
+
+		if _, err := cmd.Process.Wait(); err != nil {
+			return err
+		}
+		inst.wpipe = nil
+		inw.Close()
+		outr.Close()
+		return nil
+	}
+	go func() {
+		for inst.alive() {
+			if err := run(); err != nil && inst.debug {
+				log.Logf(0, "console: %s", err)
+			}
+			time.Sleep(1 * time.Second)
+		}
+	}()
+}
+
+func (inst *instance) alive() bool {
+	return osutil.IsExist(inst.imagePath)
+}
+
+// Run the given vmctl(8) command and wait for it to finish.
+func (inst *instance) vmctl(args ...string) ([]byte, error) {
+	if inst.debug {
+		log.Logf(0, "running command: vmctl %#v", args)
+	}
+	cmd := osutil.Command("vmctl", args...)
+	return cmd.CombinedOutput()
+}
+
+func (inst *instance) vmIdent() string {
+	return strconv.Itoa(inst.vmID)
+}
+
+// Copy the disk image since every VM needs its own disk.
+// Something similar to the snapshot feature in QEMU is not supported.
+func copyImage(src string, id int) (string, error) {
+	dirname := filepath.Dir(src)
+	basename := filepath.Base(src)
+	dst := fmt.Sprintf("%s/%s.%d", dirname, basename, id)
+	os.Remove(dst)
+	cmd := osutil.Command("cp", src, dst)
+	if err := cmd.Run(); err != nil {
+		return "", err
+	}
+	return dst, nil
+}
+
+// Extract VM ID from vmctl start output.
+func parseID(str string) (int, error) {
+	prefix := "vmctl: started vm "
+	if !strings.HasPrefix(str, prefix) {
+		return 0, fmt.Errorf("could not extract ID from: %s", str)
+	}
+	fields := strings.Fields(str)
+	i, err := strconv.Atoi(fields[3])
```

Comment:
Check that fields has at least 4 elements.
Or perhaps better:
```
strconv.Atoi(str[len(prefix):])
```
if there are no other fields.